lib/ansible/runner.rb

```diff
@@ -377,6 +378,16 @@ def python3_modules_path
       def determine_existing_python_paths_for(*paths)
         paths.select { |path| File.exist?(path) }
       end
+
+      def wait_for(dir)
+        100.times do |i|
+          break if Dir.exist?(dir)
+
+          raise "Timed out waiting for #{dir}" if i >= 99
```

Comment:
Will this ever mask a real error?  For example, is it possible that a playbook could have a bug or some other problem that prevents the artifacts from being created?  Is it always called artifacts?

Reply:
Following up on Joe's comment, maybe it shouldn't raise here and instead, log a warning and continue? That would maintain the original behavior with just the change to wait for the artifacts directory to be created

Reply:
Yeah, I agree.  Logging the unexpected situation and moving on is closer to the existing behavior and may allow playbook errors to bubble up to the logs.  I don't know that we know how to handle an exception raised here.  What do others think?

Reply:
This timeout was requested by @agrare because we did the same on the other side with the inotify implementation, so I'd ike @agrare's opinion.  My only concern is that 100 * 0.1 = 10 seconds, and on the other side we chose something like 30 seconds, so we may want to adjust the numbers to match.Even so, if we don't get an artifacts dir after 30 seconds, then the whole thing is probably broken and raising an exception seems appropriate.

Reply:
> Will this ever mask a real error? For example, is it possible that a playbook could have a bug or some other problem that prevents the artifacts from being created? Is it always called artifacts?

To clarify what ansible-runner does.  When we call ansible-runner start, it spins a child process with a pid file with an artifacts dir containing a "command" file to tell the child process what to do.  So, yes, it will _always_ have an artifacts dir _before_ we even get to the playbook execution.

Reply:
Yeah I was thinking we could keep the exact same semantics as the inotify version but just replace with a loop, but if we're not checking the `pid` file (for good reason, but still different) then that's introducing a bit of difference anyway.

We could keep the timeout (30s) the same but log instead of raising, worst case `is-alive` will return false and we'll try to load the missing artifacts dir in the response object and that will fail, but the whole thing is broken at that point anyway.

Reply:
My only concern is raising an exception is changing the behavior for a surgical backport.  The problem was that we started checking for `is-alive` before it was ready to be checked and is resolved by waiting for the artifacts directory.  

The additional raising of an exception is a nice enhancement to do on master but doesn't need to happen in this PR.  The fact that the number of seconds to wait is a concern raises warning signs to me.

Reply:
@jrafanie we do raise an exception on master though

Reply:
from side conversation:  the exception would not mask the error.  If ansible had a system level error or something that prevented the artifacts directory from being created, we would have the same chances of getting the error to be reported in ansible/system logging regardless of the exception.  The exception guarantees that we only start checking if it's `is-alive` when the artifacts directory exists.

Reply:
A few slight corrections/clarifications:

> Is it always called artifacts?

I did a very involved deep dive answering this and other questions @agrare had in my `master` version of this PR:

https://github.com/ManageIQ/manageiq/pull/20667#issuecomment-705180163

And for our purposes, "yes", since we don't change the default.

> Starting the ansible-runner process and detecting that it started should take less than a second.

Actually, I observed that this was actually "over" a second for the sub process to get started properly.  I was getting two "falses" when testing with my script above in `docker` locally without these changes, and even though they were rapid fire, it seemed liked the took a bit to of time to execute.  So it was about 2ish seconds after kicking off and returning from `ansible-runner start` that `ansible-runner is-alive` returned `true`.

Probably spawning sub-processes in containers is just slow, which caused this to happen.